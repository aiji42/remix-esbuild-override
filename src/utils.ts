export const resolve = (mod: string): string | null => {
  try {
    return require.resolve(mod);
  } catch (_) {
    return null;
  }
};

export const load = <T>(mod: string): T | null => {
  try {
    return require(mod) as T;
  } catch (_) {
    return null;
  }
};
