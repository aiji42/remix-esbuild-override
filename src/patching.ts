import * as fs from "fs";
import { resolve } from "./utils";

const pattern = /__defProps = Object\.defineProperties;/;
const redefine = `__defProp = (a, b, c) => Object.defineProperty(a, b, { ...c, configurable: true });`;

export const patching = () => {
  for (const mod of ["@remix-run/dev/node_modules/esbuild", "esbuild"]) {
    const path = resolve(mod);
    if (!path) continue;

    const original = fs.readFileSync(path, { encoding: "utf8" });
    if (original.includes(redefine)) {
      console.log("ðŸ’½ esbuild patch by remix-esbuild-override is complete.");
      break;
    }
    if (!original.match(pattern))
      throw new Error(
        "ðŸ˜µ esbuild patch by remix-esbuild-override failed, please check the the esbuild and remix versions and report this in a new issue. https://github.com/aiji42/remix-esbuild-override/issues/new"
      );
    const patched = original.replace(pattern, `$&${redefine}`);
    fs.writeFileSync(path, patched);
    console.log("ðŸ’½ esbuild patch by remix-esbuild-override is complete.");
    break;
  }
};
