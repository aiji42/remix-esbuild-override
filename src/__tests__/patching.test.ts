import { fs, vol } from "memfs";
import { patching } from "../patching";
import * as utils from "../utils";

const esbuildMainJs = `
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
`;

vi.mock("../utils");
vi.mock("fs", () => fs);
// const mockedConsoleLog = vi.spyOn(console, "log");

beforeEach(() => {
  vol.reset();
});
afterEach(() => {
  vi.resetAllMocks();
});

test("Only esbuild directly under node_modules", () => {
  vol.fromJSON(
    {
      "./node_modules/esbuild/package.json":
        '{ "name": "esbuild", "main": "lib/main.js" }',
      "./node_modules/esbuild/lib/main.js": esbuildMainJs,
    },
    "/app"
  );
  vi.spyOn(utils, "resolve").mockImplementation((mod) => {
    if (mod === "esbuild") return "/app/node_modules/esbuild/lib/main.js";
    return null;
  });

  patching();

  expect(vol.toJSON()).toMatchSnapshot();
});

test("Only esbuild under @remix-run/dev", () => {
  vol.fromJSON(
    {
      "./node_modules/@remix-run/dev/node_modules/esbuild/package.json":
        '{ "name": "esbuild", "main": "lib/main.js" }',
      "./node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js":
        esbuildMainJs,
    },
    "/app"
  );
  vi.spyOn(utils, "resolve").mockImplementation((mod) => {
    if (mod === "@remix-run/dev/node_modules/esbuild")
      return "/app/node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js";
    return null;
  });

  patching();

  expect(vol.toJSON()).toMatchSnapshot();
});

test("esbuild exists in both", () => {
  vol.fromJSON(
    {
      "./node_modules/esbuild/package.json":
        '{ "name": "esbuild", "main": "lib/main.js" }',
      "./node_modules/esbuild/lib/main.js": esbuildMainJs,
      "./node_modules/@remix-run/dev/node_modules/esbuild/package.json":
        '{ "name": "esbuild", "main": "lib/main.js" }',
      "./node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js":
        esbuildMainJs,
    },
    "/app"
  );
  vi.spyOn(utils, "resolve").mockImplementation((mod) => {
    if (mod === "@remix-run/dev/node_modules/esbuild")
      return "/app/node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js";
    return null;
  });

  patching();

  expect(vol.toJSON()).toMatchSnapshot();
});

test("defProPattern missing from script", () => {
  vol.fromJSON(
    {
      "./node_modules/esbuild/package.json":
        '{ "name": "esbuild", "main": "lib/main.js" }',
      "./node_modules/esbuild/lib/main.js": "",
    },
    "/app"
  );
  vi.spyOn(utils, "resolve").mockImplementation((mod) => {
    if (mod === "esbuild") return "/app/node_modules/esbuild/lib/main.js";
    return null;
  });

  expect(() => patching()).toThrowError(
    /esbuild patch by remix-esbuild-override failed/
  );
});
