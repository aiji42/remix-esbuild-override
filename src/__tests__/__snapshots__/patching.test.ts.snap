// Vitest Snapshot v1

exports[`Only esbuild directly under node_modules 1`] = `
{
  "/app/node_modules/esbuild/lib/main.js": "
var __create = Object.create;
var __defProp = Object.defineProperty;
__defProp = (a, b, c) => Object.defineProperty(a, b, { ...c, configurable: true });
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
",
  "/app/node_modules/esbuild/package.json": "{ \\"name\\": \\"esbuild\\", \\"main\\": \\"lib/main.js\\" }",
}
`;

exports[`Only esbuild under @remix-run/dev 1`] = `
{
  "/app/node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js": "
var __create = Object.create;
var __defProp = Object.defineProperty;
__defProp = (a, b, c) => Object.defineProperty(a, b, { ...c, configurable: true });
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
",
  "/app/node_modules/@remix-run/dev/node_modules/esbuild/package.json": "{ \\"name\\": \\"esbuild\\", \\"main\\": \\"lib/main.js\\" }",
}
`;

exports[`esbuild exists in both 1`] = `
{
  "/app/node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js": "
var __create = Object.create;
var __defProp = Object.defineProperty;
__defProp = (a, b, c) => Object.defineProperty(a, b, { ...c, configurable: true });
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
",
  "/app/node_modules/@remix-run/dev/node_modules/esbuild/package.json": "{ \\"name\\": \\"esbuild\\", \\"main\\": \\"lib/main.js\\" }",
  "/app/node_modules/esbuild/lib/main.js": "
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
",
  "/app/node_modules/esbuild/package.json": "{ \\"name\\": \\"esbuild\\", \\"main\\": \\"lib/main.js\\" }",
}
`;
